File: src/main/java/uk/co/optimisticpanda/jnomon/formatter/EventListener.java
package uk.co.optimisticpanda.jnomon.formatter;

import java.time.Duration;

import uk.co.optimisticpanda.jnomon.Utils.Colour;

public interface EventListener {
    
    default void onBeforeAll() {
        // NO-OP
    }
    
    default void onLineStart(Colour colour, Duration sinceProcessStart, Duration sinceLastStart, String line) {
        // NO-OP
    }

    default void onUpdate(Colour colour, Duration sinceProcessStart, Duration sinceLastStart) {
        // NO-OP
    }

    default void onLineEnd(Duration sinceProcessStart, Duration sinceLastStart, String lastLine) {
        // NO-OP
    }

    default void onFinally(Duration sinceProcessStart, Duration sinceLastStart) {
        // NO-OP
    }
}

File: src/main/java/uk/co/optimisticpanda/jnomon/formatter/EventListener.kt
package uk.co.optimisticpanda.jnomon.formatter

import uk.co.optimisticpanda.jnomon.Utils.Colour
import java.io.PrintStream
import java.time.Duration

interface EventListener {
  fun onBeforeAll() {
    printer().println()
  }

  fun onLineStart(colour: Colour?, sinceProcessStart: Duration, sinceLastStart: Duration, line: String?) {
    // NO-OP
  }

  fun onUpdate(colour: Colour?, sinceProcessStart: Duration, sinceLastStart: Duration) {
    // NO-OP
  }

  fun onLineEnd(sinceProcessStart: Duration, sinceLastStart: Duration, lastLine: String?) {
    // NO-OP
  }

  fun onFinally(sinceProcessStart: Duration, sinceLastStart: Duration?) {
    // NO-OP
  }

  fun printer(): PrintStream {
    return System.out
  }
}